Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp

```diff
@@ -0,0 +1,1053 @@
+            if (consumeNext() != quoteChar)
+                return didFail( HTMLFastPathResult::FailedParsingQuotedEscapedAttributeValue, UCharSpan { });
+        } else
+            return didFail(HTMLFastPathResult::FailedParsingUnquotedEscapedAttributeValue, UCharSpan { });
+        return UCharSpan { m_ucharBuffer.data(), m_ucharBuffer.size() };
+    }
+
+    void scanHTMLCharacterReference(Vector<UChar>& out)
+    {
+        ASSERT(*m_position == '&');
+        ++m_position;
+        const Char* start = m_position;
+        while (true) {
+            // A rather arbitrary constant to prevent unbounded lookahead in the case of ill-formed input.
+            constexpr int maxLength = 20;
+            if (m_position == m_end || m_position - start > maxLength || UNLIKELY(*m_position == '\0'))
+                return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+            if (consumeNext() == ';')
+                break;
+        }
+
+        CharSpan reference = CharSpan { start, static_cast<size_t>(m_position - start) - 1 };
+        // There are no valid character references shorter than that. The check protects the indexed accesses below.
+        constexpr size_t minLength = 2;
+        if (reference.size() < minLength)
+            return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+
+        if (reference[0] == '#') {
+            UChar32 res = 0;
+            if (reference[1] == 'x' || reference[1] == 'X') {
+                for (size_t i = 2; i < reference.size(); ++i) {
+                    Char c = reference[i];
+                    res *= 16;
+                    if (c >= '0' && c <= '9')
+                        res += c - '0';
+                    else if (c >= 'a' && c <= 'f')
+                        res += c - 'a' + 10;
+                    else if (c >= 'A' && c <= 'F')
+                        res += c - 'A' + 10;
+                    else
+                        return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+
+                    if (res > UCHAR_MAX_VALUE)
+                        return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+                }
+            } else {
+                for (size_t i = 1; i < reference.size(); ++i) {
+                    Char c = reference[i];
+                    res *= 10;
+                    if (c >= '0' && c <= '9')
+                        res += c - '0';
+                    else
+                        return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+
+                    if (res > UCHAR_MAX_VALUE)
+                        return didFail(HTMLFastPathResult::FailedParsingCharacterReference);
+                }
+            }
+            StringBuilder entity;
+            appendLegalEntityFor(res, entity);
+            for (unsigned i = 0; i < entity.length(); ++i)
+                out.append(entity[i]);
```

Comment:
This seems inefficient. Why not just have appendLegalEntityFor take a `Vector<UChar>`? Making and destroying a string each time won’t be fast.

Reply:
Oh good point! I'll make this change. I only used StringBuilder for consistency with the other API in that header but didn't think enough about the performance implications.

Reply:
Planning on going with:
```
void appendLegalEntityFor(UChar32 c, Vector<UChar>& decodedEntity)
{
    auto legalEntity = HTMLEntityParser::legalEntityFor(c);
    if (U_IS_BMP(legalEntity)) {
        decodedEntity.append(static_cast<UChar>(legalEntity));
        return;
    }
    decodedEntity.append(U16_LEAD(legalEntity));
    decodedEntity.append(U16_TRAIL(legalEntity));
}
```